### PR TITLE
[FLINK-33144][datastream]Deprecate Iteration API in DataStream

### PR DESCRIPTION
**Pull request by WencongLiu**
## What is the purpose of the change

Deprecate Iteration API in DataStream

## Brief change log

  - Deprecate Iteration API in DataStream

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (yes / no / don't know)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

### CODE REVIEW

**Comment by WencongLiu** [comment]
@flinkbot run azure

**Comment by WencongLiu** [comment]
@flinkbot run azure

**Comment by WencongLiu** [comment]
@flinkbot  run azure

**Comment by WencongLiu** [comment]
@flinkbot run azure

File: flink-examples/flink-examples-streaming/pom.xml [hunk 1]
```diff
@@ -137,6 +137,11 @@ under the License.
 						<configuration>
 							<compilerArgument>-Xlint:deprecation</compilerArgument>
 							<failOnWarning>true</failOnWarning>
+							<excludes>
+								<exclude>
+									org/apache/flink/streaming/examples/iteration/IterateExample.java
```
Comment (xintongsong) [inline flink-examples/flink-examples-streaming/pom.xml:142]: We should explain that this example is temporarily preserved only for testing purpose.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/datastream/DataStream.java [hunk 2]
```diff
@@ -526,8 +526,14 @@ public DataStream<T> global() {
      * in the set time, the stream terminates.
      *
      * @return The iterative data stream created.
+     * @deprecated This method is deprecated since Flink 1.19. The users are recommended to use
+     *     Iteration API in Flink ML instead.
```
Comment (xintongsong) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/datastream/DataStream.java:530]: We should not recommend users to use Iteration API in Flink ML instead. It doesn't make sense that a user who doesn't need any ML algorithm would have to use Flink ML only to get access to the Iteration API.

I'd suggest to the following:
> The only known use case of this Iteration API comes from Flink ML, which already has its own implementation of iteration and  no longer uses this API. If there's any use cases other than Flink ML that needs iteration support, please reach out to dev@flink.apache.org and we can consider making the Flink ML iteration implementation a separate common library.

**Review by xintongsong — APPROVED** [review]
LGTM

**Pull request closed without merging**